vlasovsolver/cpu_trans_pencils.hpp

```diff
@@ -97,6 +106,85 @@ struct setOfPencils {
       path.push_back(pathIn);
    }
 
+   void binPencils() {
+      binOfPencil.resize(N);
+
+      // Consider only cells which _any_ pencil writes into for binning,
+      // since read-only cells aren't affected by race conditions
+      std::unordered_set<CellID> targetCells = {};
+      #pragma omp parallel for
+      for (uint i = 0; i < sumOfLengths; ++i) {
+         const CellID targ = ids[i];
+         const Realf ratio = targetRatios[i];
+         if (targ && (ratio > 0.0)) {
+            #pragma omp critical
+            targetCells.insert(targ);
+         }
+      }
+
+      // Loop over pencils
+      // TODO could be paralellized as well
+      for (uint i = 0; i < N; ++i) {
+         binOfPencil[i] = i;
+         targetCellsInBin[i] = {};
+
+         for (auto id = ids.begin() + idsStart[i]; id < ids.begin() + idsStart[i] + lengthOfPencils[i]; ++id) {
+            // We don't need to consider source and target cells of the pencil separately
+            // as all pencils with source/target cell C must be in the same bin as all pencils with target C
+            if (*id && targetCells.contains(*id)) {
+               targetCellsInBin[i].insert(*id);
+            }
+         }
+      }
+
+      // Super ugly!
+      // TODO If anyone knows how to do this more efficiently feel free to fix it
+      std::set<uint> binsToDelete;
+      do {
+         binsToDelete.clear();
+         for (auto bin1 = targetCellsInBin.begin(); bin1 != targetCellsInBin.end(); ++bin1) {
```

Comment:
could we use e.g. `[binIndex1, targetCellsInBin1]` so instead of `bin1->first` below, we'd have `binIndex`?